clang-tools-extra/clang-tidy/bugprone/VirtualArithmeticCheck.cpp

```diff
@@ -0,0 +1,49 @@
+//===--- VirtualArithmeticCheck.cpp - clang-tidy---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "VirtualArithmeticCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void VirtualArithmeticCheck::registerMatchers(MatchFinder *Finder) {
+  const auto PointerExprWithVirtualMethod =
+      expr(hasType(pointerType(pointee(hasDeclaration(
+               cxxRecordDecl(hasMethod(isVirtualAsWritten())))))))
+          .bind("pointer");
+
+  const auto ArraySubscript =
+      arraySubscriptExpr(hasBase(PointerExprWithVirtualMethod));
+
+  const auto BinaryOperators =
+      binaryOperator(hasAnyOperatorName("+", "-", "+=", "-="),
+                     hasEitherOperand(PointerExprWithVirtualMethod));
+
+  const auto UnaryOperators =
+      unaryOperator(hasAnyOperatorName("++", "--"),
+                    hasUnaryOperand(PointerExprWithVirtualMethod));
+
+  Finder->addMatcher(
+      expr(anyOf(ArraySubscript, BinaryOperators, UnaryOperators)), this);
```

Comment:
I added the type alias matches as requested, but the pure virtual functions matcher has the same limitation as the regular virtual functions matcher - it cannot transitively decide if a function is pure or not, so it's quite limited as well. Added tests and a FIXME for it.

Reply:
Creating this matcher
```c++
AST_MATCHER(CXXRecordDecl, isAbstract) { return Node.isAbstract(); }
```
will provide a way to transitively know if a record is abstract, i.e. if any pure virtual function was not overridden, which is a guaranteed way to find out if it should be considered polymorphic. The matcher can simply be added inside an `anyOf` together with the `hasMethod` in both pointer expr matchers (`anyOf(isAbstract(), hasMethod(...))`). (I realize I didn't even mention `isAbstract` in the above comment in the inheritance part, which was the whole reason for that section)

Also, please move the `unless(isFinal())` matcher before the `hasMethod` matcher for slightly better performance.

Reply:
Added `isAbstract`, and fixed `unless(isFinal())` as well (although with the amount of `final` classes in my test projects, it might even be slower in practice...).

Reply:
> with the amount of final classes in my test projects, it might even be slower in practice...

Hm, good point.

I ran this check with `time run-clang-tidy -clang-tidy-binary ./bin/clang-tidy -checks="-*,*polymorphic*" CodeGen` and couldn't measure any real difference when swapping around any of the matchers inside the `cxxRecordDecl` around. So maybe this isn't needed. I was originally about to suggest moving `isAbstract` before `hasMethod`, but that didn't matter either, at least in llvm. If you measure a noticeable difference in another project, then you can probably adjust the order of the matchers.